Hashmap for vrf lookup by table

nathan-duddles opened the pull request:
### Summary
Add hashmap for lookup of vrf_id by table_id to `vrf_lookup_by_id()`.

`vrf_lookup_by_id()` moved from `zebra/rt_netlink.c` to `lib/vrf.c`

### Related Issue
Add Mapping From Table ID to VRF #5982

### Components
zebra, lib

nathan-duddles committed:
zebra: Remove vrf_lookup_by_table decl/def from rt_netlink

vrf_lookup_by_table() function is used outside of rt_netlink.c
currently in if_netlink.c and may be useful in other files
in the future

Will add to vrf.c[h], as this currently is home to other vrf lookup
functions

Signed-off-by: Nathan Duddles <duddles.nathan@gmail.com>

nathan-duddles committed:
lib: Add vrf_lookup_by_table with hashmap to vrf.c[h]

Lookup uses hashmap with table_id key and vrf_id values.
Only for vrf backended by vrf lite. Could do similar
implementation with ns_id in future.

Hashmap local to vrf.c but must be updated when
vrf is added or modified. For updates to vrf_id,
call to helper function vrf_ids_by_table_update()
is added in the publicly used vrf_update() function.

Previously, the table_id was set by directly modifying
vrf->data.l.table_id in zclient.c. However, as no functions
in vrf.c are notified when change occurs, a helper function,
vrf_update_table_id(), is added. This both updates
vrf_data.l.table_id, and updates the local hashmap.

vrf_delete() function used for removing vrfs. Add call to
here to vrf_ids_by_table_delete() helper function
to remove entries from hashmap.

Signed-off-by: Nathan Duddles <duddles.nathan@gmail.com>

idryzhov reviewed (changes requested):
I don't see any reason to make this part of the library. `vrf_lookup_by_table` is only ever used in zebra.

nathan-duddles commented:
Sorry, I'm not sure how to interpret these test failures. It looks like all TopoTests Ubuntu 18.04 i386 failed. 

Running topotests locally in an Ubuntu 20.04 x86_64 VM, I'm only getting a single failing test, which occurs when run on both FRRouting:master  and nathan-duddles:hashmap_for_vrf_lookup_by_table. Failing test is:
ospf_basic_functionality/test_ospf_chaos.py::test_ospf_chaos_tc34_p1

Also, it like some warnings were generated for formatting (e.g. trailing white spaces). I did see a linter test failing when initially creating the pull request and I made the recommended changes, amended the commit and force pushed. I guess these warnings reflect the code before those changes.

nathan-duddles commented:
Thank you for the comment @idryzhov . I had moved it to `lib/vrf.c[h]` based on the recommendation to do so in the GH issue: https://github.com/FRRouting/frr/issues/5982 but I am new to contributing to FRR, so I will defer to others where it is best located. I'll just note that `lib/vrf.c[h] is where similar functions are defined: `vrf_lookup_by_id`, `vrf_lookup_by_name`, `vrf_id_to_name`.

I believe it was originally added in `zebra/rt_netlink.c` because it is used there. It is also used in `zebra/if_netlink.c`. As I mentioned in the comments of issue, I think this function could also be called at [pimd/pim_iface.c at line 1714](https://github.com/FRRouting/frr/blob/964240dab9fe6a7347d05c92ed97b1f6ecf60f71/pimd/pim_iface.c#L1714), where lookup of vrf by table_id is currently done using the inefficient RB_FOREACH method.

donaldsharp commented:
bgpd is crashing.  Here is some valgrind hints as to what is going wrong:

==884032== Conditional jump or move depends on uninitialised value(s)
==884032==    at 0x4991A76: vrf_ids_by_table_const_find (vrf.c:83)
==884032==    by 0x4991AEE: vrf_ids_by_table_find (vrf.c:83)
==884032==    by 0x4991C6D: vrf_ids_by_table_delete (vrf.c:92)
==884032==    by 0x4992369: vrf_delete (vrf.c:361)
==884032==    by 0x4992A94: vrf_terminate_single (vrf.c:675)
==884032==    by 0x4992B83: vrf_terminate (vrf.c:703)
==884032==    by 0x1EB633: bgp_vrf_terminate (bgp_main.c:362)
==884032==    by 0x1EB20C: bgp_exit (bgp_main.c:253)
==884032==    by 0x1EB020: sigint (bgp_main.c:170)
==884032==    by 0x4974A74: frr_sigevent_process (sigevent.c:130)
==884032==    by 0x498DA98: thread_fetch (thread.c:1771)
==884032==    by 0x49242CF: frr_run (libfrr.c:1195)
==884032==    by 0x1EBB00: main (bgp_main.c:519)
==884032==    
==884032== Conditional jump or move depends on uninitialised value(s)
==884032==    at 0x4991AC3: vrf_ids_by_table_const_find (vrf.c:83)
==884032==    by 0x4991AEE: vrf_ids_by_table_find (vrf.c:83)
==884032==    by 0x4991C6D: vrf_ids_by_table_delete (vrf.c:92)
==884032==    by 0x4992369: vrf_delete (vrf.c:361)
==884032==    by 0x4992A94: vrf_terminate_single (vrf.c:675)
==884032==    by 0x4992B83: vrf_terminate (vrf.c:703)
==884032==    by 0x1EB633: bgp_vrf_terminate (bgp_main.c:362)
==884032==    by 0x1EB20C: bgp_exit (bgp_main.c:253)
==884032==    by 0x1EB020: sigint (bgp_main.c:170)
==884032==    by 0x4974A74: frr_sigevent_process (sigevent.c:130) 
==884032==    by 0x498DA98: thread_fetch (thread.c:1771)
==884032==    by 0x49242CF: frr_run (libfrr.c:1195)
==884032==    by 0x1EBB00: main (bgp_main.c:519)
==884032==

nathan-duddles committed:
lib: Fix memory issues with hash map

Each entry in hash map must be allocated memory
on the heap.

In addition, much check that head is initialized
prior to calling _add() and _del() functions.

Also made corrections to compare and hash functions.

Fix formatting causing build warnings

Signed-off-by: Nathan Duddles <duddles.nathan@gmail.com>

nathan-duddles commented:
Thank you providing the hint and the Valgrind output!

It looks like I failed to call `vrf_ids_by_table_init(&vrf_ids_by_table_head)`. I have added a flag/check on whether or not it is initialized, so that it is initialized the first time `vrf_ids_by_table_update()` is called. If there is a better way to do this, please do let me know!

As I'm not getting this failing case locally, is it possible to push changes to this branch and run the CI check again? Shall I amend and force push to the same commit?

sworleys commented:
> I don't see any reason to make this part of the library. `vrf_lookup_by_table` is only ever used in zebra.

while currently only being used by zebra, the function makes sense to made general for VRFs and added to lib in my opinion. The API could be useful for other daemons. Pbr is one example that deals with both vrfs/table's directly.

sworleys commented on lib/vrf.c:
```diff
@@ -57,6 +58,62 @@ static int vrf_backend;
+
+static uint32_t vrf_ids_by_table_hash(const struct vrf_ids_by_table_proxy *item)
+{
+	return item->table_id;
```
even though those table_id's should be unique, I think you still need to call `jhash_1word()` here to keep the table sparse.

nathan-duddles replied:
Yes, I agree. I looked for other uses of jhash_1word() in the codebase and followed suit. I generated and hard coded a random initval, which seems to be how it was done elsewhere.

sworleys commented:
> As I'm not getting this failing case locally, is it possible to push changes to this branch and run the CI check again? Shall I amend and force push to the same commit?

yes you can force-push and re-run CI. Do it sparingly though as it is a shared resource.

nathan-duddles commented on lib/vrf.c:
```diff
@@ -57,6 +58,62 @@ static int vrf_backend;
+static int vrf_ids_by_table_cmp(const struct vrf_ids_by_table_proxy *item_a,
+				const struct vrf_ids_by_table_proxy *item_b)
+{
+	return (item_a->table_id == item_b->table_id);
```
Good point! I changed this to return 1 for a>b, -1 for a<b and 0 for a==b.

nathan-duddles commented:
Sorry, I didn't realize the issue with explicitly initializing a static variable to zero. Seems this was flagged as an ERROR by the OpenBSD 7 amd64. I've corrected this, along with other WARNINGS related to formatting and re-pushed.

nathan-duddles commented:
This patch failed two tests for Ubuntu 18.04 arm8:
https://ci1.netdef.org/build/result/viewBuildResultsFailedTests.action?buildKey=FRR-PULLREQ2-TOPO6U18ARM8&buildNumber=4243
Which did pass on Ubuntu 18.04 amd64:
https://ci1.netdef.org/browse/FRR-PULLREQ2-TOPO6U18AMD64-4243/test/case/109021784
https://ci1.netdef.org/browse/FRR-PULLREQ2-TOPO6U18AMD64-4243/test/case/109021787

I'm sorry I haven't been able to figure out why these be failing on Ubuntu 18.04 arm8. I only have access to an amd64 machine and running all topotests on Ubuntu 18.04 is passing for me locally. (Running with valgrind shows possible memory leaks in unrelated modules and these same warnings appear when running on main without this patch.)

If anyone has any ideas why these tests may be failing or suggestions on how to debug this, I would really appreciate any feedback. Thanks!

donaldsharp commented:
those tests are not unknown to occassionally fail.  Let me rerun

donaldsharp commented:
ci:rerun snmp test fialure?

nathan-duddles commented:
Ok, thanks! I also just noticed from the warning that there is a comment to myself that I should have removed. I'll fix this by amending the last commit.

sworleys commented:
@nathan-duddles looks mostly good to me. I think the way you are checking for initialization is fine.

could you address the checkpatch warning?

you can recreate this locally by turning your patches into a single patch file, reverting the tree to before they are applied and doing:
./tools/checkpatch.sh ../my-patch.txt ./

it's the same script the linux kernel uses for style checking

```
Report for vrf.c | 2 issues
===============================================
< WARNING: Missing a blank line after declarations
< #161: FILE: /tmp/f1-18129/vrf.c:161:
```

idryzhov commented on lib/vrf.c:
```diff
@@ -20,6 +20,7 @@
  */
 
 #include <zebra.h>
+#include "zebra/zebra_vrf.h"
```
This is not right. We can't include zebra-specific headers in the library file.

nathan-duddles replied:
Thank you for catching this. I've removed it in the most recent commit.

idryzhov commented on lib/vrf.c:
```diff
@@ -279,6 +401,50 @@ struct vrf *vrf_lookup_by_id(vrf_id_t vrf_id)
+	/* case vrf with netns : match the netnsid */
+	if (vrf_is_backend_netns()) {
+		RB_FOREACH (vrf, vrf_id_head, &vrfs_by_id) {
+			zvrf = vrf->info;
```
This code is only correct for zebra.
`vrf->info` can not be casted to `struct zebra_vrf` in other daemons.

nathan-duddles replied:
I've refactored this in the most recent commit. The cast to `zebra_vrf` stuct is not needed here and is not indicative of daemon-specific functionality.

idryzhov reviewed (changes requested):
Once again, the functionality is daemon-specific.
And the current implementation only confirms this.
We can't use daemon-specific structs and headers in the library code.

sworleys commented:
> Once again, the functionality is daemon-specific. And the current implementation only confirms this. We can't use daemon-specific structs and headers in the library code.

Hmm... yea @idryzhov is correct here. I didn't realize netns-backed vrfs breaks it from being shared knowledge in lib.

Let's just move the code to zebra/zebra_vrf.c[h] for now probably and leave it zebra specific to get the code merged.

nathan-duddles commented:
Thank you both for your review and comments on this! Good catch @idryzhov on the inclusion of zebra_vrf. 

Actually, `zebra_vrf` functionality isn't needed; as I wasn't adding a hash method for netns-backed vrf, I just avoiding modify that, but I should have since moving it into vrf.c. The only zebra_vrf function used was `zvrf_id()`, which is simply returning `vrf->vrf_id` for the encapsulated vrf struct. I've done the refactor and removed the include of `zebra_vrf` and use of `zvrf_id()`. Am testing locally to make sure this still works and then will push a new commit.

nathan-duddles committed:
lib: Refactor to remove use of zebra_vrf

Use of zebra_vrf was from when vrf_lookup_by_table()
was defined in rt_netlink. Use of zebra_vrf function:
zvrf_id() is not needed, as this just encapsulates
the return of vrf->vrf_id.

Also fixed formatting to pass checkpath warning
on line 161.

Signed-off-by: Nathan Duddles <duddles.nathan@gmail.com>

nathan-duddles commented:
When you have time, please take a look at the updates I've made and let me know if more changes are needed.

As I mentioned above, I think `vrf_lookup_by_table()` could also be called at [pimd/pim_iface.c at line 1714](https://github.com/FRRouting/frr/blob/964240dab9fe6a7347d05c92ed97b1f6ecf60f71/pimd/pim_iface.c#L1714), where lookup of vrf by table_id is currently done using the inefficient RB_FOREACH method.

If `vrf_lookup_by_table()` can be added to `vrf.c` in this pull request, I can make the update to `pim_iface` and submit another pull request for that.

nathan-duddles commented:
Hello! The autoclose tag reminded me about this pull request. I would love to get this reviewed and merged if you are still wanting the patch. Please let me know if there is anything else I can do on this. Thanks!

riw777 commented:
@frrbot autoclose in 1 month
